controlplane/kubeadm/controllers/controller.go

```diff
@@ -263,6 +264,18 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 		return ctrl.Result{}, nil
 	}
 
+	// All infrastructure machines created by KCP have infra-template ref annotation, but the ones that are adopted (that are not created by a machine set) or created before this change do not.
+	// Infra machines that doesn't have infra-template ref annotations are being patched here.
+	// This needs to be called after adoptMachines and before MachinesNeedingUpgrade() to avoid unnecessary upgrade
+	// on the newly adopted machines and older version machines.
```

Comment:
Can we avoid this reconciliation with the semantic equality work (https://github.com/kubernetes-sigs/cluster-api/issues/3170)? Otherwise I worry that we might be misrepresenting where the template was cloned from as part of this.

Reply:
This was actually done for that work, basically we need to backfill the machines that don't have the annotation yet so we can compare the external references later on. We can do semantic equality on the CABPK embedded types, but not necessarily on the infrastructure and bootstrap references (given that they'd be unstructured). 

We'll have to put a big warning in place that before upgrading to v0.3.7 folks need to make sure any KCP upgrades have been done

Reply:
Might be I'm missing something, but why not to set infra ref as part of the adoption process vs having a separated reconcile loop for newly adopted machines

Reply:
It's not just adoption, the backfill is needed for normal, pre-existing machines as well

Reply:
I'm wondering if we can get away with only adding the annotation for net new resources and somehow ignore the absence for pre-existing resources? That would avoid the potential to backfill with incorrect information, and I think as long as we are clear about it in the release notes, there shouldn't be any user confusion around behaviors.

Reply:
Would we be able to detect a reference change later on if we ignore it? 🤔

Reply:
If we ignore the absence, then we would only be able to detect a reference change after an upgrade has taken place (for pre-existing infra).

Reply:
Right, would we then require for folks to upgrade the Kubernetes version if they only want to change the infrastructure reference?

Reply:
They wouldn't necessarily need to update the kubernetes version, they could also trigger an "upgrade" using the UpgradeAfter field...  I think my concern is that we could end up setting the annotation and then when downstream controllers are consuming it we could potentially trigger mutations to the underlying infrastructure that is not expected/wanted/applicable. This could be something minor like modifying instance tags or could potentially end up causing errors if there are changes that would be considered immutable (instance type differs between linked template and the existing resource) that could have undefined behavior as we allow for providers to reconcile template changes down to instantiated resources.

Reply:
That sounds good to me if that's the behavior we want 👍 

cc @ncdc

Reply:
Also, for adopted machines, if we don't backfill, all adopted external machines will be missing the annotation which will make it an ongoing thing in the future releases too.

If an external machine doesn't have annotation, just checking infra ref change during MachinesNeedingUpgrade() would be enough?

Reply:
> Also, for adopted machines, if we don't backfill, all adopted external machines will be missing the annotation which will make it an ongoing thing in the future releases too.

Indeed, at least for the extent that we wish to support adoption. Not sure if we've adopted a stance on dropping support for adoption for v1alpha4, or if we would want to drop support for a later release (my preference would be dropping at v1alpha4)

I would, however, argue this is likely appropriate behavior, since we cannot say for sure that the template matches the existing resource during adoption (at least not without having some method of doing a semantic comparison between the two, but that would all be provider specific).

> If an external machine doesn't have annotation, just checking infra ref change during MachinesNeedingUpgrade() would be enough?

Yes, I believe so.

That said, I believe the idea with allowing for providers to configure a subset of templates to be mutable was to allow for the ability to make certain changes in place without the need for triggering an upgrade through KCP (such as applying tags, security groups, or other similar changes that don't require re-provisioning the instance).